### PR TITLE
metrics for monitoring QRep and CDC flows along with Prometheus and Grafana docker containers

### PR DESCRIPTION
**Pull request by heavycrystal**

### CODE REVIEW

File: flow/connectors/bigquery/bigquery.go [hunk 1]
```diff
@@ -667,6 +669,7 @@ func (c *BigQueryConnector) NormalizeRecords(req *model.NormalizeRecordsRequest)
 	if err != nil {
 		return nil, fmt.Errorf("failed to execute statements %s in a transaction: %v", strings.Join(stmts, "\n"), err)
 	}
+	c.logNormalizeMetrics(req.FlowJobName, distinctTableNames)
```
[the remark this answers is not in the document]
Comment (heavycrystal) [reply flow/connectors/bigquery/bigquery.go:672]: fixed

File: flow/connectors/postgres/client.go [hunk 2]
```diff
@@ -428,3 +426,53 @@ func (c *PostgresConnector) generateUpdateStatement(allCols []string, unchangedT
...
+	countTablesResults := c.pool.SendBatch(c.ctx, countTablesBatch)
+	err := countTablesResults.Close()
+	if err != nil {
+		log.Errorf("error while closing statement batch: %v", err)
+		return 0, fmt.Errorf("error while closing statement batch: %w", err)
+	}
+	return totalCount, nil
+}
+
+func (c *PostgresConnector) getMaxWaterMarkColumn(tableName string, watermarkColumn string) (int64, error) {
```
[the remark this answers is not in the document]
Comment (heavycrystal) [reply flow/connectors/postgres/client.go:458]: removed for now

File: flow/connectors/snowflake/snowflake.go [hunk 3]
```diff
@@ -549,15 +552,26 @@ func (c *SnowflakeConnector) NormalizeRecords(req *model.NormalizeRecordsRequest
...
 			tableNametoUnchangedToastCols[destinationTableName],
 			getRawTableIdentifier(req.FlowJobName),
 			syncBatchID, normalizeBatchID, normalizeRecordsTx)
 		if err != nil {
 			return nil, err
 		}
+		totalRowsAffected += rowsAffected
+	}
+	if totalRowsAffected > 0 {
+		err = c.logNormalizeMetrics(req.FlowJobName, int64(totalRowsAffected), time.Since(startTime),
```
[the remark this answers is not in the document]
Comment (heavycrystal) [reply flow/connectors/snowflake/snowflake.go:570]: fixed

File: flow/connectors/snowflake/qrep_avro_sync.go [hunk 4]
```diff
@@ -50,10 +50,13 @@ func (s *SnowflakeAvroSyncMethod) SyncQRepRecords(
 
 	stage := s.connector.getStageNameForJob(config.FlowJobName)
 
+	putFileStartTime := time.Now()
 	err = s.putFileToStage(localFilePath, stage)
 	if err != nil {
 		return 0, err
 	}
+	s.connector.logQRepSyncMetrics(config.FlowJobName, int64(len(records.Records)),
```
[the remark this answers is not in the document]
Comment (heavycrystal) [reply flow/connectors/snowflake/qrep_avro_sync.go:58]: fixed

File: flow/connectors/snowflake/client.go [hunk 5]
```diff
@@ -63,3 +63,22 @@ func NewSnowflakeClient(ctx context.Context, config *protos.SnowflakeConfig) (*S
...
 }
+
+func (c *SnowflakeConnector) getTableCounts(tables []string) (int64, error) {
+	var totalRecords int64
+	for _, table := range tables {
+		_, err := parseTableName(table)
+		if err != nil {
+			return 0, fmt.Errorf("failed to parse table name %s: %w", table, err)
+		}
+		query := fmt.Sprintf("SELECT COUNT(*) FROM %s", table)
```
[the remark this answers is not in the document]
Comment (heavycrystal) [reply flow/connectors/snowflake/client.go:74]: necessary, adding //nolint:gosec

File: docker-compose.yml [hunk 6]
```diff
@@ -77,20 +77,65 @@ services:
       temporal-admin-tools:
         condition: service_healthy
 
-  flow_worker:
-    container_name: flow_worker
+  flow_worker1:
```
Comment (iskakaushik) [inline docker-compose.yml:80]: create a new file - `multi-worker.docker-compose.yml`, leave the current one as minimal.

File: flow/connectors/postgres/client.go [hunk 7]
```diff
@@ -428,3 +425,31 @@ func (c *PostgresConnector) generateUpdateStatement(allCols []string, unchangedT
 	}
 	return strings.Join(updateStmts, "\n")
 }
+func (c *PostgresConnector) getTableCounts(tables []string) (int64, error) {
```
Comment (iskakaushik) [inline flow/connectors/postgres/client.go:428]: @heavycrystal  replace this with count estimates: https://wiki.postgresql.org/wiki/Count_estimate, counts can be really slow for some of the tables we are mirroring.

**Review by iskakaushik — APPROVED** [review]

**Pull request merged**